_posts/2018-03-26-proxstar.md

```diff
@@ -0,0 +1,27 @@
+author: Jordan Rodgers
+author-image: https://avatars1.githubusercontent.com/u/5183608?s=460&v=4
+author-bio: Fourth Year Networking & Systems Administration Student at Rochester Institute of Technology
+author-email: com6056@csh.rit.edu
+author-social:
+  github: https://github.com/com6056
+---
+
+## The Problem
+CSH currently uses Proxmox for VM hosting. It has been working great for everything we use it work with one exception: member VM self-service. The interface isn't very intuitive to someone who might not have created a VM before, creating a higher barrier to entry than we would like for members who might want to start learning how to use Linux for the first time. It also lacked the ability to enforce resource limits and automatically cleanup VMs that members weren't using anymore. All of my previous major projects were much more ops and/or security focused so I figured writing something to solve this problem would be a great way to do a much more development focused one and expand my developer skillset.
+
+## The Solution
+Proxstar (a mashup of Proxmox and STARRS) is a web application that allows for easy VM self-service for the members of CSH. It is written in Python using Flask and runs on the CSH OpenShift cluster. RQ is being used as a task queue for tasks that take more than a second or two and periodic tasks (RQ-Scheduler). Proxstar allowed members to create, edit, and delete VMs and will automatically handle any network registration that needs to be done in STARRS for them. During creation, they can specify to either create it from a template or to install it themselves from an ISO. If creating it from a template, Proxstar will spin up a ready-to-go VM with a a user that has a randomly generated password for the member to use.
+
+#### Making RTP Lives Easier
+Having to periodically go through and clean up VMs that haven't been touched by members in months can be a tedious task. Enforcing usage limits can also be difficult since Proxmox doesn't have a mechanism for it. Proxstar allows RTPs to set a limit on the number of CPU cores, memory, and disk space a member can utilize. It also sets an expiration date on each new VM, emails the member before it expires, and automatically deletes any VMs that aren't renewed.
+
+#### Struggles
+The hardest part of the project was figuring out how to give console access to members. Proxmox does not provide a friendly way to access their VNC consoles, so I had to dig into it to figure out how they did it and try to do a similar thing within Proxmox. Upon clicking the 'Console' button, Proxstar generates a valid token for the target that will be created with noVNC, forwards the port that the VNC session will be spun up on from the respective Proxmox host to the VNC instance of Proxstar, starts the VNC session for the VM on Proxmox, and opens a new tab with the noVNC session. It isn't the cleanest solution, but it has held up pretty well.
```

Comment:
`s/within Proxmox/within Proxstar/`, I assume.

Reply:
Clarify what the "target" is here - I assume it's the target VM, but say so explicitly. This sentence is also a bit too long and confusing; please try to break it up a bit more.